### PR TITLE
[SPARK-38700][SQL] Use error classes in the execution errors of save mode

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Migrate the following errors in QueryExecutionErrors:
 
* unsupportedSaveModeError -> UNSUPPORTED_SAVE_MODE

### Why are the changes needed?
Porting execution errors of unsupported saveMode to new error framework.
 
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new UT.

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -196,6 +196,9 @@
   "UNSUPPORTED_OPERATION" : {
     "message" : [ "The operation is not supported: <operation>" ]
   },
+  "UNSUPPORTED_SAVE_MODE" : {
+    "message" : [ "The saveMode is not supported: <saveMode> (<pathExists>)" ]
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:200]: The message is not clear from user side, I believe. Let's introduce two sub-classes EXISTENT_PATH and NON_EXISTENT_PATH. Messages should be something like:
1. The save mode <saveMode> is not supported for: an existent path.
2. The save mode <saveMode> is not supported for: a not existent path.

What's does true or false in the error message mean is not clear from my point of view.

See https://github.com/apache/spark/pull/36307 for an example.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 2]
```diff
@@ -424,4 +424,17 @@ class QueryExecutionErrorsSuite
       matchMsg = true
     )
   }
+
+  test("UNSUPPORTED_SAVE_MODE: unsupported null saveMode") {
+    withTempPath { path =>
+      val e = intercept[SparkIllegalStateException] {
+        val saveMode: SaveMode = null
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:431]: Besides of passing a null, is it possible to trigger the error from user space?

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:437]: SaveMode only have four value,
https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/SaveMode.java#L27~L57
Append, Overwrite, ErrorIfExists, Ignore

According to the following code, only can trigger QueryExecutionErrors.unsupportedSaveModeError when SaveMode = null.
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L114~L138

File: core/src/main/scala/org/apache/spark/ErrorInfo.scala [hunk 3]
```diff
@@ -80,9 +80,9 @@ private[spark] object SparkThrowableHelper {
       val errorSubInfo = subClass.getOrElse(subErrorClass,
         throw new IllegalArgumentException(s"Cannot find sub error class '$subErrorClass'"))
       val subMessageParameters = messageParameters.tail
-      "[" + errorClass + "." + subErrorClass + "] " + errorInfo.messageFormat +
-        String.format(errorSubInfo.messageFormat.replaceAll("<[a-zA-Z0-9_-]+>", "%s"),
-          subMessageParameters: _*)
+      "[" + errorClass + "." + subErrorClass + "] " + String.format((errorInfo.messageFormat +
+        errorSubInfo.messageFormat).replaceAll("<[a-zA-Z0-9_-]+>", "%s"),
+        subMessageParameters: _*)
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/ErrorInfo.scala:85]: Could you explain why do you need to modify this, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 4]
```diff
@@ -587,8 +587,13 @@ object QueryExecutionErrors extends QueryErrorsBase {
        """.stripMargin)
   }
 
-  def unsupportedSaveModeError(saveMode: String, pathExists: Boolean): Throwable = {
-    new IllegalStateException(s"unsupported save mode $saveMode ($pathExists)")
+  def saveModeUnsupportedError(saveMode: Any, pathExists: Boolean): Throwable = {
+    pathExists match {
+      case true => new SparkIllegalStateException(errorClass = "UNSUPPORTED_SAVE_MODE",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:592]: I think it is better to use `SparkIllegalArgumentException` here.

File: core/src/main/scala/org/apache/spark/ErrorInfo.scala (hunk 3 above)
Comment (panbingkun) [reply core/src/main/scala/org/apache/spark/ErrorInfo.scala:85]: 1.
"UNSUPPORTED_SAVE_MODE" : {
    "message" : [ "The save mode <saveMode> is not supported for: " ],
    "subClass" : {
      "EXISTENT_PATH" : {
        "message" : [ "an existent path." ]
      },
      "NON_EXISTENT_PATH" : {
        "message" : [ "a not existent path." ]
      }
    }
  },

2.
https://github.com/panbingkun/spark/blob/master/core/src/main/scala/org/apache/spark/ErrorInfo.scala#L82~L91

3.
The placeHolder(eg. <saveMode>) in ErrorInfo's messageFormat need replace.
I think this is also necessary in other scenarios.

**Comment by MaxGekk** [comment]
@panbingkun Could you resolve conflicts, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve conflicts, please.

Done, Thanks @MaxGekk

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Comment by MaxGekk** [comment]
@panbingkun Could you backport the changes to branch-3.3, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you backport the changes to branch-3.3, please.

i just see the message today, sorry!

**Pull request closed without merging**